testsuite/integration-arquillian/servers/auth-server/services/testsuite-providers/src/main/java/org/keycloak/testsuite/forms/SetClientNoteAuthenticator.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.keycloak.testsuite.forms;
+
+import java.util.List;
+
+import javax.ws.rs.core.MultivaluedMap;
+
+import org.jboss.logging.Logger;
+import org.keycloak.Config;
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.authentication.Authenticator;
+import org.keycloak.authentication.AuthenticatorFactory;
+import org.keycloak.models.AuthenticationExecutionModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.provider.ProviderConfigProperty;
+import org.keycloak.sessions.AuthenticationSessionModel;
+
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class SetClientNoteAuthenticator implements Authenticator, AuthenticatorFactory {
+
+    protected static final Logger logger = Logger.getLogger(SetClientNoteAuthenticator.class);
+
+    public static final String PROVIDER_ID = "set-client-note-authenticator";
+
+    // Query parameters of this name will be used to save the client note to authentication session
+    public static final String PREFIX = "note-";
+
+    @Override
+    public void authenticate(AuthenticationFlowContext context) {
+        MultivaluedMap<String, String> inputData = context.getHttpRequest().getDecodedFormParameters();
```

Comment:
Do we really want to push whatever comes from the request to the session notes? Does not look safe for me ...

I think that ideally, we should have some switch to enable/disable this on a per-flow basis or have a `Transient Session Authenticator` that always set this note to indicate whether or not sessions for that flow should be transient.

Reply:
Or maybe this is something that should be hardcoded for now into Docker protocol which is the initial problem we want to solve.

Reply:
@pedroigor Sure, this is not safe. This is testsuite-only authenticator, which is not intended to be ever used in production. I've used authenticator just to be able to test this for the OAuth2 DirectGrant, as the Docker protocol automated tests don't work.

As you pointed, in case of the docker protocol, it is currently hardcoded - the note is set in the DockerEndpoint class. Other protocols (OIDC, SAML) currently don't have support for transient session in this PR - with the only exception of the testsuite, which have this SetClientNoteAuthenticator .

Do you think that we need authenticator to be added to the main codebase to be able to have transient sessions per flow? I can think that OAuth2 Resource Owner Password Credentials Grant (Direct Grant) in some cases doesn't need user session at all. AFAIK Sometimes people use DirectGrant just for the single authentication to some CLI app and never need to refresh or introspect tokens etc. Not sure if this is useful also for some other protocols - like for example SAML ECP? WDYT?

Reply:
Ohh, sorry. Didn't notice it was related to the test suite.

But yeah, I'm OK with having something in the main codebase to achieve that. Not sure if the authenticator approach is the best path though ... At the same time, I'm not sure if adding a config option to authentication flows is something we want to do.

SAML ECP is a good example, yeah.

Btw, transient sessions should not work if later you try to introspect the token using the token introspection endpoint, right? I mean, people should be aware that the lifetime of tokens issued using a transient session is bound to their time-related claims and can/should only be validated locally by the resource server. And no support for revocation but wait until the time-related claims become invalid.

Reply:
> Btw, transient sessions should not work if later you try to introspect the token using the token introspection endpoint, right?

That's a good point, there should be a docs update in https://www.keycloak.org/docs/latest/server_admin/#_docker

The PR looks good to me as is once the docs are in place. My proposal would be to merge this PR as is once the docs are ready, and include no authenticator or other means to set the client note in the main code base now. It is straightforward to mark the session to be transient, and the best means to do so may depend on the particular use case. @pedroigor WDYT?

Reply:
@hmlnarik @pedroigor Well, I am starting to think that Authenticator approach may be nice and will have some advantages :) Administrator can decide if he wants to create transient session by simply adding the authenticator to the authentication flow. May be useful for DirectGrant or SAML ECP. For Docker protocol, we can have this authenticator in the flow by default, but admin will have a chance to remove it if he wants and have sessions persistent for docker as well for some reason.

Authenticatior will have some additional flexibility that administrator may want to persist sessions just under specified condition (hence the authenticator can be added to some conditional subflow and be triggered just in case of the condition passed)... This is something, which won't be easily available if we have this as hardcoded option of the authentication flow.

On the other hand, this won't be classic "Authenticator", but rather some kind of interceptor in the authentication flow. But our Authentication SPI don't have this concept, so just use it as implementation of "Authenticator" interface can be possible.

So my ideal way would be to change this PR to have the Authenticator approach, but if you prefer, I can keep it as is and just update the documentation. WDYT?

I agree that docs will need to be updated. As pointed, transient session would mean that introspection, "refresh token" or "Revoke session" won't be possible. I will update the docs once we refine whether to go with authenticator approach or keep the PR as is.

Reply:
The authenticator approach it is certainly interesting and possible way to deal with it, yet I am not convinced yet it's the right way. For example, there is no separate SAML / ECP flow so it would not be possible to use an authenticator in this case.

At this moment, I'd proceed with this PR as is, and only follow up with a common approach once we have it.

Reply:
I agree. I'm still not 100% sure about the Authenticator but is what we have ... @mposolda I agree this looks like a new concept (or, potentially, a new configuration option for Authentication Flows).

In any case, LGTM. Perhaps we should make this somehow tech-preview and advertize that it may change in the future.

Reply:
@hmlnarik @pedroigor Ok, Thanks. Let's wait with the authenticator or something else to later if needed. The docs PR is here https://github.com/keycloak/keycloak-documentation/pull/1045 .

I did not know SAML ECP profile does not have separate flow, Thanks for clarifying. I somehow assumed that yes based on the fact that Direct grant (OAuth2 Resource Owner PAssword Credentials) has it and I though this is somehow similar.